### ISSUE TITLE
Feature request: Angular AoT compilation without bundling

### ISSUE THREAD

**Issue by kasperhhk** [angular/angular#12390]
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report => search github for a similar issue or PR before submitting
[x ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Currently we are forced by AoT to compile all of the javascript into 1 bundle, that is then optimized for size. This makes sense in a classical web scenario.

**Please tell us about your environment:**
We are running a server with HTTP/2, and we are running a site that employs multiple angular applications per page. (for example 1 app for the menu and 1 app for a form)

**Please tell us about the requested feature:**
Currently (in development) we are using SystemJS to load in modules, one by one.
However the overhead on starting the app up is extreme (around 1000ms minimum), so we were thinking that AoT would allow us to compile the views, so we could drop the angular-compiler completely and thereby gain a huge performance and size reduction.

However AoT is currently only setup to do bundling, with rollup only bundling 1 entry-point to 1 file.

So either we hack an entry point that includes ALL of our apps and we get 1 gigantic file with lots of unused code per pageload, or we make a rollup-config per app and setup a build that builds each app in turn, giving us a bundle per app, which means we're going to have a LOT of redundant code since the library code will be included in each bundle.

What we'd like is the possibility of simply compiling the templates and components and other typescript files and then load them in as needed. Thus removing the need for a runtime transpiler, but still keeping to the modular nature of typescript.

(Personally i tried getting the ngfactory files themselves to work with Systemjs, but that's a never-ending battle to even get it to work on simple things, since it suddenly wants to use the sourcefiles for angular components, instead of the .umd.js bundles. Meaning we'd have to transpile the sourcefiles at runtime, which is dumb and counterproductive to the whole point of AoT.)

Any comments or information that could help is of course welcomed.
- **Angular version:** 2.1.0
- **Node (for AoT issues):** `node --version` =

**Comment by robwormald** [comment]
There's nothing inherent in AoT that *requires* you to build a bundle - you could absolutely load the files one by one (though even with http2, loading hundreds of modules is ill-advised...)

I suggest having a look at https://github.com/systemjs/builder#example---arithmetic-expressions - this would allow you to bundle vendor code into one System.register bundle, and retain the ability to bring in individual features.

**Comment by kasperhhk** [comment]
Except i've tried loading the AoT files one by one (loading a main file that bootstraps an ngfactory module), and it doesn't work. It starts requesting ngfactory files from node_modules/@angular/core/src, and then transpiling those because they aren't ES5.

**Comment by kasperhhk** [comment]
Interesting with the systemjs builder, that would be very handy if we could get AoT to work without rollup. I'll look into it a bit further 👍

Also saying that it is ill-advised to load hundreds of modules with http2 is all well and fine, but i'd like to at least *try*, so that i can see what kind of performance we get with and without bundling, how much we need to bundle and so on. It's the fact that i've been unable to even make it work that's the focus.

**Comment by kasperhhk** [comment]
When i tried to use systemjs-builder, it didn't transpile the views, so we're back to square one, since ngfactory bootstrapped angular doesn't work with systemjs.

**Comment by abakirov** [comment]
Hi,

> There's nothing inherent in AoT that requires you to build a bundle

AoT compilation creates `*.ngfoctory.ts` files with references to angular src folder like this:

```
import * as import1 from '@angular/core/src/linker/view';
```

Which implies rollup bundling step after AoT compilation. Is there a way to maybe configure `ngc` to generate ngfactory files with imports like the following (instead of the above)?

```
import { AppView } from '@angular/core';
```

This would allow us to use Angular umd bundles with AoT-compiled code to accommodate the use case described in this issue.

**Comment by mlc-mlapis** [comment]
I am also still missing something principal. What I want to do is to load by SystemJS loader equivalently the same modules as in JIT but now translated by ngc to AOT. It means ES5 code (no templates and no compiler in a browser). But now there are references to @angular src folders in compiled *.ngfactory.ts files generated by ngc (they are not ES5 files, they are TS/ES2015 files).

As I understand there should not be differences in source code to be able to run an app in JIT and AOT mode. 

Like the code:
```
this._loader.load('./app/modules/machinery/machinery.module#MachineryModule')
```
when "_loader" is injected through DI in the constructor and it relates to SystemJsNgModuleLoader class -> instantiated as a sigleton provider during app init.

This should load file "machinery.module.js" in JIT and "machinery.module.ngfactory.js" in AOT.

**Comment by AbhilashThomas** [comment]
Hey, any update on this?

**Comment by abakirov** [comment]
with changes done in #14388 (and included into 4.0.0-beta.8 and later), AoT now generates shallow imports like this:
```
import * as import0 from '@angular/core';
```
so this works now, may thanks!

**Comment by mfreeman-xtivia** [comment]
I too would love to see a recommended answer for this. There does not currently seem to be any good answer for the case where you have multiple pages containing Angular 2 apps and would like to ....

1. Use AoT to compile the application parts, bundle each app, and then have these bundled AoT apps....
2. each reference a shared bundled version of the A2 core libraries that could be leveraged across all of the pages

I'd prefer to do this some using  the SystemJS loader, but as was previously noted the fact that ngc generates bucket of low-level imports like

`import * as import1 from '@angular/core/src/linker/view';`

makes this seem un-achievable at the moment?

**Comment by sbabeal** [comment]
We are running into the same problem with the last ng2 compiler.  Would be nice to know why this is happening and what to do about it.  I guess the ngfactory files and the node_module files could be merged for the compiled modules and then map system.js to this folder but that's a lot of work.

**Comment by IgorMinar** [comment]
this particular issue has been resolved already in v4.0 by changing the generated imports to point to `@angular/core` rather than `@angular/core/some/internal/path`. It's should now be trivial to setup systemjs configuration that works as you described above.

if there is another issue that prevents the usecases discussed above, please file a new issue. thanks!

**Comment by mlc-mlapis** [comment]
@IgorMinar @sbabeal ... I can confirm that it is possible to compile any Angular 5.x app using just NGC and use SystemJS without any deep linking problems.
